flang/lib/Optimizer/Transforms/OpenACCDataOperandConversion.cpp

```diff
@@ -0,0 +1,198 @@
+    Location loc = op.getLoc();
+    fir::LLVMTypeConverter &converter = *static_cast<fir::LLVMTypeConverter *>(
+        OpConversionPattern<Op>::getTypeConverter());
+    unsigned numDataOperand = op.getNumDataOperands();
+
+    // Keep the non data operands without modification.
+    auto nonDataOperands =
+        operands.take_front(operands.size() - numDataOperand);
+    llvm::SmallVector<mlir::Value> convertedOperands;
+    convertedOperands.append(nonDataOperands.begin(), nonDataOperands.end());
+
+    // Go over the data operand and legalize them for translation.
+    for (unsigned idx = 0; idx < numDataOperand; ++idx) {
+      mlir::Value originalDataOperand = op.getDataOperand(idx);
+
+      //
+      if (auto refTy =
+              originalDataOperand.getType().dyn_cast<fir::ReferenceType>()) {
+        if (refTy.getEleTy().isa<fir::SequenceType>() ||
+            fir::isa_std_type(refTy.getEleTy())) {
+          // Basic arrays and scalars are passed as llvm.ptr to the translation.
+          // Code to compute the size will be generated during the translation.
+          auto convertedType = converter.convertType(refTy);
+          auto convertedValue = builder.create<fir::CastOp>(
+              loc, convertedType, originalDataOperand);
+          convertedOperands.push_back(convertedValue);
+        } else if (auto boxTy = refTy.getEleTy().dyn_cast<fir::BoxType>()) {
+          // BoxType needs more work to extract the correct information.
+          return builder.notifyMatchFailure(
+              op, "fir.box type currently not supported");
+
+          // When a BoxType is encountered we need to push the descriptor and
+          // the actual data.
+
+          auto convertedType =
+              converter.convertType(refTy).cast<mlir::LLVM::LLVMPointerType>();
+          auto convertedValue =
+              builder
+                  .create<fir::CastOp>(loc, convertedType, originalDataOperand)
+                  .res();
+
+          // TODO: currently this code is not doing what it should just so dummy
+          // code to create and populate the data descriptor with info.
+          auto fortranDescriptor =
+              builder.create<LLVM::LoadOp>(loc, convertedValue);
+          LLVM::LLVMStructType fortranDescrType =
+              convertedType.getElementType().dyn_cast<LLVM::LLVMStructType>();
+          auto descrElementPtr = builder.create<LLVM::ExtractValueOp>(
+              loc, fortranDescrType, fortranDescriptor,
+              builder.getI32ArrayAttr(0));
+
+          auto descr =
+              DataDescriptor::undef(builder, loc, fortranDescrType.getBody()[0],
+                                    fortranDescrType.getBody()[0]);
+          descr.setBasePointer(builder, loc, descrElementPtr);
+          descr.setPointer(builder, loc, descrElementPtr);
+          // Compute size?
+          // descr.setSize(builder, loc, descrElementSize);
+
+          convertedOperands.push_back(descr);
```

Comment:
So if I understand things correctly, right now the code is only pushing a `DataDescriptor` for the data itself, but the plan is to also push a  `DataDescriptor` to the fortran descriptor ? What's the benefit/purpose of having both rather than only the  a `DataDescriptor` to the fortan descriptor ?

Reply:
We can actually push only one `DataDescriptor` with the base pointer being the fortran descriptor and the second pointer being the pointer to the data. The runtime can handle this case.